### PR TITLE
New: SFW. Disabling SFW when the database was outdated.

### PR DESCRIPTION
**Pull request by Glomberg**

### CODE REVIEW

File: cleantalk.php [hunk 1]
```diff
@@ -638,6 +638,23 @@ function apbct_sfw__check()
         return;
     }
 
+    // Checking if database was outdated
+    if ( $apbct->stats['sfw']['last_update_time'] + $apbct->stats['sfw']['update_period'] * 3 < time() ) {
```
Comment (safronik) [inline cleantalk.php:642]: There is a method State::errorToggle() it could perfectly fit here.

File: lib/Cleantalk/ApbctWP/State.php [hunk 2]
```diff
@@ -669,6 +669,11 @@ public function errorToggle($add_error, $type, $error, $major_type = null, $set_
         }
     }
 
+    public function errorExists($error_type)
```
Comment (safronik) [inline lib/Cleantalk/ApbctWP/State.php:674]: Maybe method is unnecessary

File: lib/Cleantalk/ApbctWP/State.php [hunk 3]
```diff
@@ -294,7 +294,9 @@ class State
             'last_send_time'          => 0,
             'last_send_amount'        => 0,
             'last_update_time'        => 0,
+            'last_update_way'         => '',
```
Comment (safronik) [inline lib/Cleantalk/ApbctWP/State.php:297]: Unused flag

**Review by safronik — CHANGES_REQUESTED** [review]

File: cleantalk.php (hunk 1 above)
Comment (Glomberg) [reply cleantalk.php:642]: State::errorToggle() method does not check the existence of the error and every time rewrite this. It is not good for the performance.

File: lib/Cleantalk/ApbctWP/State.php (hunk 2 above)
Comment (Glomberg) [reply lib/Cleantalk/ApbctWP/State.php:674]: See the previous comment

File: lib/Cleantalk/ApbctWP/State.php (hunk 3 above)
Comment (Glomberg) [reply lib/Cleantalk/ApbctWP/State.php:297]: The flag is using here https://github.com/CleanTalk/wordpress-antispam/blob/sfw_shutdown/cleantalk.php#L1285

File: cleantalk.php (hunk 1 above)
Comment (safronik) [reply cleantalk.php:642]: You could extend the method to use it this way.

Comment (Glomberg) [reply cleantalk.php:642]: Done https://github.com/CleanTalk/wordpress-antispam/pull/40/commits/4d7dc2151e32e65b1bd083ec9221cbdd8ddc7767

**Review by safronik — APPROVED** [review]

**Pull request merged**